[FLINK-7668] Add ExecutionGraphCache for ExecutionGraph based REST handlers

tillrohrmann opened the pull request:
## What is the purpose of the change

The ExecutionGraphCache replaces the ExecutionGraphHolder. Unlike the latter, the former
does not expect the AccessExecutionGraph to be the true ExecutionGraph. Instead it assumes
it to be the ArchivedExecutionGraph. Therefore, it invalidates the cache entries after
a given time to live period. This will trigger requesting the AccessExecutionGraph again
and, thus, updating the ExecutionGraph information for the ExecutionGraph based REST
handlers.

In order to avoid memory leaks, the WebRuntimeMonitor starts now a periodic cleanup task
which triggers ExecutionGraphCache.cleanup. This methods releases all cache entries which
have exceeded their time to live. Currently it is set to 20 * refreshInterval of the
web gui.

## Brief change log

- Rename `ExecutionGraphHolder` into `ExecutionGraphCache`
- Make `ExecutionGraphCache` thread safe by using a `ConcurrentHashMap`
- Add `ExecutionGraph` request logic depending on the time to live for each cache entry
- Add periodic cleanup task to `WebRuntimeMonitor` which triggers `ExecutionGraphCache#cleanup`

## Verifying this change

This change added tests and can be verified as follows:

- `ExecutionGraphCacheTest` adds multiple tests

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

zentol commented on flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/WebRuntimeMonitor.java:
```diff
@@ -228,55 +245,55 @@ public WebRuntimeMonitor(
 		Router router = new Router();
 		// config how to interact with this web server
-		get(router, new DashboardConfigHandler(executor, cfg.getRefreshInterval()));
+		get(router, new DashboardConfigHandler(scheduledExecutor, cfg.getRefreshInterval()));
```
well i see an easy way to reduce the diff...

tillrohrmann replied:
But it's not necessarily about reducing the diff if the name change makes it clearer what this object is.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMasterGateway.java:
```diff
@@ -227,5 +229,19 @@ void failSlot(final ResourceID taskManagerId,
+	 * @param timeout for the rpc call
+	 * @return Future details of the executed job
+	 */
 	CompletableFuture<JobDetails> requestJobDetails(@RpcTimeout Time timeout);
```
are we still keeping this for the job master?

tillrohrmann replied:
`JobDetails` are independent of the `ExecutionGraphCache` and still needed. I simply added java docs.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/RestfulGateway.java:
```diff
@@ -38,5 +44,36 @@
+	 * @param timeout for the asynchronous operation
+	 * @return Future containing the AccessExecutionGraph for the given jobId, otherwise {@link JobNotFoundException}
+	 */
+	CompletableFuture<AccessExecutionGraph> requestJob(JobID jobId, @RpcTimeout Time timeout);
```
Well this seems questionable. These methods technically aren't related to being a restful gateway,

This seems like an odd attempt to reduce code duplication by just moving the code one layer up without other considerations.

tillrohrmann replied:
It makes sense to accumulate (all) methods required for the REST handlers in a common interface because it will be implemented by the `Dispatcher` as well as the `JobMaster`. Right now, the `RestfulInterface` is used two purposes: 1. retrieving the REST address and 2. all REST handler related requests. Once all information which is required by the REST handlers is serializable, we no longer have to redirect. Then we can drop the `requestRestAddress` method. Then I would also rename this interface into something like `RestInformationGateway`. Until then, I would like to keep it how it is.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/ExecutionGraphCache.java:
```diff
@@ -0,0 +1,182 @@
+		private final CompletableFuture<AccessExecutionGraph> executionGraphFuture;
+
+		ExecutionGraphEntry(long ttl) {
+
```
odd empty line

tillrohrmann replied:
I think this up to personal taste.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/ExecutionGraphCache.java:
```diff
@@ -0,0 +1,182 @@
+
+			final ExecutionGraphEntry newEntry = new ExecutionGraphEntry(currentTime + timeToLive.toMilliseconds());
+
+			final boolean success;
```
lets tgive this a more meaningful name

tillrohrmann replied:
Will do

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/ExecutionGraphCache.java:
```diff
@@ -0,0 +1,182 @@
+			}
+
+			if (success) {
+				// request job graph
```
unnecessary

tillrohrmann replied:
true

zentol commented on flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/legacy/ExecutionGraphCacheTest.java:
```diff
@@ -0,0 +1,357 @@
+				CompletableFuture<AccessExecutionGraph> executionGraphFuture = CompletableFuture
+					.supplyAsync(
+						() -> executionGraphCache.getExecutionGraph(jobId, jobManagerGateway),
+						TestingUtils.defaultExecutor())
```
isn't the default executor single-threaded?

tillrohrmann replied:
This is a good point. Will change it.

zentol commented on flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/legacy/ExecutionGraphCacheTest.java:
```diff
@@ -0,0 +1,357 @@
+
+			assertEquals(toBeSuspendedExecutionGraph, executionGraphFuture.get());
+
+			// switch to suspended
```
unnecessary

tillrohrmann replied:
True

zentol commented on flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/WebRuntimeMonitor.java:
```diff
@@ -193,11 +198,23 @@ public WebRuntimeMonitor(
 		}
 
-		ExecutionGraphHolder currentGraphs = new ExecutionGraphHolder(timeout);
+		final long timeToLive = cfg.getRefreshInterval() * 10L;
```
the commit message says it is refreshInterval * 20.

tillrohrmann replied:
yes, because `10 * 2 = 20`.

zentol replied:
ah right, the message was referring to the cleanupInterval, my bad.

tillrohrmann commented:
Thanks for the review @zentol. I've addressed your comments.

Since the `WebRuntimeMonitor` requests the actual `ExecutionGraph` from the `JobManagerGateway` instead of the `ArchivedExecutionGraphs`, there need not be a low refresh interval. It will be updated instantaneously.

zentol commented:
But that's not the case for the flip-6 code, is it? There the JobMaster creates a separate ArchivedExecutionGraph that is not automatically updated.

tzulitai commented on flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/legacy/ExecutionGraphCacheTest.java:
```diff
@@ -0,0 +1,357 @@
+
+			CompletableFuture<AccessExecutionGraph> executionGraphFuture2 = executionGraphCache.getExecutionGraph(jobId, jobManagerGateway);
+
+			assertEquals(accessExecutionGraph, executionGraphFuture2.get());
```
Maybe we should also verify that the gateway is requested two times `requestJob` here also, to check that a cache entry was not created for the first failing request.

tillrohrmann replied:
The former is checked implicitly by returning the `accessExecutionGraph`.

Not sure if we should test the component with such a level of detail. It basically involves a lot of implementation knowledge. From the user perspective it should not make a difference if we create a failed entry which is updated or no entry. The important bit is that it will retry it.

tzulitai replied:
I see. Ok, lets leave it as is then.

tzulitai commented on flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/legacy/ExecutionGraphCacheTest.java:
```diff
@@ -0,0 +1,357 @@
+
+			Thread.sleep(timeToLive.toMilliseconds());
+
+			executionGraphCache.cleanup();
```
I'm not sure how this test differentiates from `testExecutionGraphEntryInvalidation`.
Not sure if I'm correct, but the only difference is this `cleanup()` call. So, it could well be that this test can also pass because of the TTL sleep, while `cleanup()` can be no-op.

Maybe we need to expose the internal entries of the cache just for tests, so that we can verify entry cleanup without invoking `getExecutionGraph`

tillrohrmann replied:
You're right. I'll add a `size()` method to solve the problem.

tzulitai commented on flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/legacy/ExecutionGraphCacheTest.java:
```diff
@@ -0,0 +1,357 @@
+		when(suspendedExecutionGraph.getState()).thenReturn(JobStatus.SUSPENDED);
+
+		final JobManagerGateway jobManagerGateway = mock(JobManagerGateway.class);
+		// let's first answer with a JobNotFoundException and then only with the correct result
```
invalid comment "let's first answer with a JobNotFoundException"
-> "lets first answer with a suspended ExecutionGraph"

tillrohrmann replied:
Good catch. Will change it.

tzulitai commented on flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/legacy/ExecutionGraphCacheTest.java:
```diff
@@ -0,0 +1,357 @@
+
+			CompletableFuture<AccessExecutionGraph> executionGraphFuture2 = executionGraphCache.getExecutionGraph(jobId, jobManagerGateway);
+
+			assertEquals(accessExecutionGraph, executionGraphFuture2.get());
```
Need to verify `jobManagerGateway#requestJob` is issued 2 times (to assert cache invalidation).

tillrohrmann replied:
I think this is implicitly tested by having different `AccessExecutionGraphs` returned by the different calls.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/ExecutionGraphCache.java:
```diff
@@ -0,0 +1,180 @@
+				if (currentTime < oldEntry.getTTL()) {
+					if (oldEntry.getExecutionGraphFuture().isDone() && !oldEntry.getExecutionGraphFuture().isCompletedExceptionally()) {
+
+						// TODO: Remove once we no longer request the actual ExecutionGraph from the JobManager but only the ArchivedExecutionGraph
```
yes

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/ExecutionGraphCache.java:
```diff
@@ -0,0 +1,182 @@
+							}
+							// send a new request to get the ExecutionGraph from the new leader
+						} catch (InterruptedException | ExecutionException e) {
+							throw new RuntimeException("Could not retrieve ExecutionGraph from the orderly completed exception. This should never happen.", e);
```
Actually it should be `orderly completed future`. Good catch.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/JobConfigHandler.java:
```diff
@@ -42,7 +42,7 @@
 	private static final String JOB_CONFIG_REST_PATH = "/jobs/:jobid/config";
 
-	public JobConfigHandler(ExecutionGraphHolder executionGraphHolder, Executor executor) {
+	public JobConfigHandler(ExecutionGraphCache executionGraphHolder, Executor executor) {
```
Indeed. Will change it.

tillrohrmann commented:
Thanks for the review @zentol and @tzulitai.

@zentol yes you are right. The Flip-6 `JobMaster` will only return the `ArchivedExecutionGraph`. I intend to set the `timeToLive` for the `ExecutionGraphCache` to the `refreshInterval` of the web ui.

tillrohrmann commented:
I've rebased the code onto the latest master. Are all PR comments resolved @zentol and @tzulitai?

tzulitai commented:
Thanks for addressing / replying to my comments.
LGTM, +1 from me

zentol commented:
+1 from my side too.

tillrohrmann commented:
Thanks for your review @zentol and @tzulitai. Merging this PR.
